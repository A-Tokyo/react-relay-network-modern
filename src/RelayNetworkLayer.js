/* @flow */

import { Network } from 'relay-runtime';
import RelayRequest from './RelayRequest';
import fetchWithMiddleware from './fetchWithMiddleware';
import type {
  Middleware,
  FetchFunction,
  SubscribeFunction,
  RNLExecuteFunction,
} from './definition';

export default class RelayNetworkLayer {
  _middlewares: Middleware[];
  execute: RNLExecuteFunction;
  fetchFn: FetchFunction;
  subscribeFn: ?SubscribeFunction;

  constructor(middlewares: Middleware[] | Middleware, subscribeFn?: SubscribeFunction) {
    this._middlewares = Array.isArray(middlewares) ? middlewares : [middlewares];
    this.subscribeFn = subscribeFn;

    this.fetchFn = (operation, variables, cacheConfig, uploadables) => {
      const req = new RelayRequest(operation, variables, cacheConfig, uploadables);
      return fetchWithMiddleware(req, this._middlewares);
    };

    this.execute = Network.create(this.fetchFn, this.subscribeFn);
  }
}
