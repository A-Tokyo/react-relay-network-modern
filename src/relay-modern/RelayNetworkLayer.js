/* @flow */

import { Network } from 'relay-runtime';
import type { Middleware } from '../definition';

export type Variables = { [name: string]: $FlowFixMe };
export type ConcreteBatch = any;
export type CacheConfig = {
  force?: ?boolean,
  poll?: ?number,
  rerunParamExperimental?: ?any,
};
export type Disposable = { dispose(): void };
export type Uploadable = File | Blob;
export type UploadableMap = { [key: string]: Uploadable };
export type PayloadData = { [key: string]: mixed };
export type QueryPayload = {|
  data?: ?PayloadData,
  errors?: Array<any>,
  rerunVariables?: Variables,
|};
// this is workaround should be class from relay-runtime/network/RelayObservable.js
export type RelayObservable<T> = Promise<T>;
// Note: This should accept Subscribable<T> instead of RelayObservable<T>,
// however Flow cannot yet distinguish it from T.
export type ObservableFromValue<T> = RelayObservable<T> | Promise<T> | T;
export type FetchFunction = (
  operation: ConcreteBatch,
  variables: Variables,
  cacheConfig: CacheConfig,
  uploadables: ?UploadableMap
) => ObservableFromValue<QueryPayload>;
// See SubscribeFunction type declaration in relay-runtime/network/RelayNetworkTypes.js
export type SubscribeFunction = (
  operation: ConcreteBatch,
  variables: Variables,
  cacheConfig: CacheConfig,
  observer: any
) => RelayObservable<QueryPayload> | Disposable;

export default class RelayModernNetworkLayer {
  _middlewares: Middleware[];
  execute: (
    operation: ConcreteBatch,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables?: ?UploadableMap
  ) => RelayObservable<QueryPayload>;

  constructor(middlewares: Middleware[] | Middleware, subscribeFn?: SubscribeFunction) {
    this._middlewares = Array.isArray(middlewares) ? middlewares : [middlewares];

    const fetchFn: FetchFunction = (operation, variables, cacheConfig, uploadables) => {
      return Promise.resolve({});
    };

    this.execute = Network.create(fetchFn, subscribeFn);
  }
}
