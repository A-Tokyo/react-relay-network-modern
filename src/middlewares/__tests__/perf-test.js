/* @flow */

import fetchMock from 'fetch-mock';
import RelayNetworkLayer from '../../RelayNetworkLayer';
import { mockReq } from '../../__mocks__/mockReq';
import perfMiddleware from '../perf';
import batchMiddleware from '../batch';

describe('middlewares/perf', () => {
  beforeEach(() => {
    fetchMock.restore();
  });

  it('measure request time for single request', async () => {
    fetchMock.mock({
      matcher: '/graphql',
      response: {
        status: 200,
        body: { data: 'PAYLOAD' },
      },
      method: 'POST',
    });

    const logger = jest.fn();
    const rnl = new RelayNetworkLayer([
      perfMiddleware({
        logger,
      }),
    ]);

    await mockReq('MyRequest').execute(rnl);
    expect(logger).toHaveBeenCalledTimes(1);
    expect(logger.mock.calls[0][0]).toMatch(/\[\d+ms\] MyRequest/);
    expect(logger.mock.calls[0][1]).toMatchSnapshot();
  });

  it('measure request time for batch request', async () => {
    fetchMock.mock({
      matcher: '/graphql/batch',
      response: {
        status: 200,
        body: [{ id: 1, data: 'PAYLOAD' }, { id: 2, data: 'PAYLOAD' }],
      },
      method: 'POST',
    });

    const logger = jest.fn();
    const rnl = new RelayNetworkLayer([
      batchMiddleware(),
      perfMiddleware({
        logger,
      }),
    ]);

    mockReq(1).execute(rnl);
    await mockReq(2).execute(rnl);
    expect(logger).toHaveBeenCalledTimes(1);
    expect(logger.mock.calls[0][0]).toMatch(/\[\d+ms\] BATCH_REQUEST:1:2/);
    expect(logger.mock.calls[0][1]).toMatchSnapshot();
  });
});
