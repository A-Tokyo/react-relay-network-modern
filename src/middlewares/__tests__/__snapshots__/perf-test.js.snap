// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middlewares/perf measure request time for request 1`] = `
RelayRequest {
  "cacheConfig": Object {},
  "fetchOpts": Object {
    "body": "{\\"id\\":\\"MyRequest\\",\\"query\\":\\"\\",\\"variables\\":{}}",
    "headers": Object {},
    "method": "POST",
  },
  "id": "MyRequest",
  "operation": Object {
    "id": "MyRequest",
    "text": "",
  },
  "uploadables": undefined,
  "variables": Object {},
}
`;
